cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,193 @@
+		AutoAllocateBlocks: false,
+		StrictAffinity:     true,
+	}); err != nil {
+		log.WithError(err).Fatal("failed to disable auto-allocation of block")
+	}
+
+	// Read in all the files in the host-local directory.
+	files, err := ioutil.ReadDir(ipAllocPath)
+	if err != nil {
+		log.WithField("path", ipAllocPath).WithError(err).Fatal("failed to read path")
+	}
+
+	// For each file, convert it into an IP allocation and then delete the file.
+	for _, f := range files {
+		fname := path.Join(ipAllocPath, f.Name())
+
+		// Delete and skip the last reserved IP.
+		if strings.Contains(f.Name(), "last") {
+			if err := os.Remove(fname); err != nil {
+				log.WithField("file", fname).WithError(err).Fatal("failed to remove file")
+			}
+			continue
+		}
+
+		// The name of the file is its IP address.
+		ip, _, err := cnet.ParseCIDR(fmt.Sprintf("%s/32", f.Name()))
+		if err != nil {
+			log.WithField("file", f.Name()).WithError(err).Fatal("failed to parse CIDR")
+		}
+
+		// The contents are the container ID.
+		b, err := ioutil.ReadFile(fname)
+		if err != nil {
+			log.WithField("file", fname).WithError(err).Fatal("failed to read file")
+		}
+
+		// Store allocation to datastore
+		err = assignCalicoIP(calicoClient, *ip, string(b), nodename)
+		if err != nil {
+			log.WithField("nodename", nodename).WithError(err).Fatal("failed to assign IP to calico backend")
+		}
+
+		// Delete the file
+		if err := os.Remove(fname); err != nil {
+			log.WithField("file", fname).WithError(err).Fatal("failed to remove file")
+		}
+	}
+
+	// Delete the directory
+	if err := os.Remove(ipAllocPath); err != nil {
+		log.WithField("path", ipAllocPath).WithError(err).Fatal("failed to remove directory")
+	}
+
+	// Check DaemonSet rolling update status to see if this is the last node to update.
+	// If so, re-enable auto-allocation of blocks
+	k8sCfg, err := clientcmd.BuildConfigFromFlags("", "")
+	if err != nil {
+		log.WithError(err).Fatal("failed to get kubeconfig")
+	}
+	k8sClient, err := kubernetes.NewForConfig(k8sCfg)
+	if err != nil {
+		log.WithError(err).Fatal("failed to initialize kube client")
+	}
+	resp, err := k8sClient.ExtensionsV1beta1().DaemonSets("kube-system").Get("calico-node", v1.GetOptions{})
+	if err != nil {
+		log.WithError(err).Fatal("failed to get daemonsets/kube-system.calico-node")
+	}
+
+	// UpdatedNumberScheduled tracks the total number of nodes that are running updated daemon pod.
+	if resp.Status.UpdatedNumberScheduled == resp.Status.DesiredNumberScheduled-1 {
+		if err = calicoClient.IPAM().SetIPAMConfig(context.Background(), ipam.IPAMConfig{
+			AutoAllocateBlocks: true,
+			StrictAffinity:     false,
+		}); err != nil {
+			log.WithError(err).Fatal("failed to re-enable auto-allocation of block")
+		}
+	}
+}
+
+func assignCalicoIP(c client.Interface, ip cnet.IP, containerID, nodename string) error {
+	handleID := fmt.Sprintf("%s.%s", "k8s-pod-network", containerID)
+	return c.IPAM().AssignIP(context.Background(), ipam.AssignIPArgs{
```

Comment:
Ah, this is an interesting one... this PR adds some attributes: https://github.com/projectcalico/cni-plugin/pull/680

Which, the cleanup controller relies on. We should figure out what to do in this code. Options include:
- Query the apiserver to figure out the pod this IP belongs to
- Add some other attribute the allocation cleanup controller can use to clean this up.
- ???


We should at least include the node in the attributes though (as the linked PR does).